src/layouts/hexbin.js

```diff
@@ -0,0 +1,112 @@
+        pi = Math.round(px = px - (pj & 1) / 2),
+        py1 = py - pj;
+    if (Math.abs(py1) * 3 > 1) {
+      let px1 = px - pi,
+        pi2 = pi + (px < pi ? -1 : 1) / 2,
+        pj2 = pj + (py < pj ? -1 : 1),
+        px2 = px - pi2,
+        py2 = py - pj2;
+      if (px1 * px1 + py1 * py1 > px2 * px2 + py2 * py2) pi = pi2 + (pj & 1 ? 1 : -1) / 2, pj = pj2;
+    }
+    const key = `${pi}|${pj}`;
+    keys.set(key, [pi, pj]);
+    return key;
+  })
+  .filter(([p]) => p)
+  .map(([p, bin]) => {
+    const [pi, pj] = keys.get(p);
+    bin.x = (pi + (pj & 1) / 2) * dx;
+    bin.y = pj * dy;
+    return bin;
+  });
+}
+
+// Allow hexbin options to be specified as part of outputs; merge them into options.
+function mergeOptions({radius = 10, ...outputs}, options) {
+  return [outputs, {radius, ...options}];
+}
+
+function hexbinLayout(radius, outputs, options) {
+  // we defer to Plot.bin’s reducers, but some of them are not supported
+  for (const reduce of Object.values(outputs)) {
+    if (typeof reduce === "string"
+    && !reduce.match(/^(first|last|count|distinct|sum|deviation|min|min-index|max|max-index|mean|median|variance|mode|proportion|proportion-facet)$/i))
+      throw new Error(`invalid reduce ${reduce}`);
+  }
+  outputs = maybeOutputs(outputs, options);
+  const rescales = {
+    r: {scale: "r", options: {range: [0, radius * w0]}},
+    fill: {scale: "color"},
+    stroke: {scale: "color"},
+    fillOpacity: {scale: "opacity"},
+    strokeOpacity: {scale: "opacity"},
+    symbol: {scale: "symbol"}
+  };
+  const {x, y} = options;
+  if (x == null) throw new Error("missing channel: x");
+  if (y == null) throw new Error("missing channel: y");
+  return layout({...defaults, ...options}, function(index, scales, {x: X, y: Y}) {
+    const values = {x: [], y: [], r: []};
+    const channels = [];
+    const newIndex = [];
+    for (const o of outputs) {
+      o.initialize(this.data);
+      o.scope("data", index);
+    }
+    let n = 0;
+    for (const I of index) {
+      const facetIndex = [];
+      newIndex.push(facetIndex);
+      const bins = hbin(I, X, Y, radius);
+      for (const o of outputs) {
+        o.scope("facet", I);
+        for (const bin of bins) o.reduce(bin);
+      }
+      for (const bin of bins) {
+        values.x[n] = bin.x;
+        values.y[n] = bin.y;
+        facetIndex.push(n++);
+      }
+    }
+    for (const o of outputs) {
+      if (o.name in rescales) {
+        const {scale, options} = rescales[o.name];
+        const value = o.output.transform();
+        channels.push([o.name, {scale, value, options}]);
+      } else {
+        values[o.name] = o.output.transform();
+      }
+    }
+    if (!channels.find(([key]) => key === "r")) values.r = Array.from(values.x).fill(radius);
+
+    return {index: newIndex, values, channels};
+  });
+}
+
+export function hexbin(outputs, options) {
+  ([outputs, options] = mergeOptions(outputs, options));
+  const {radius, ...inputs} = options;
```

Comment:
For both hexbin and hexgrid, I think we’ll want to use *r* instead of *radius* for consistency with dot.

Reply:
Actually, it’s complicated. We can’t just use *r* because you might want to use the *r* channel, e.g.,

```js
Plot.hexbin({r: "sum"}, {x: "culmen_depth_mm", y: "culmen_length_mm", r: "body_mass_g"})
```

so you need a separate option to specify the size of the hexagonal grid.

At the same time, though, it would really be nice to say:

```js
Plot.hexbin({fill: "count"}, {x: "culmen_depth_mm", y: "culmen_length_mm", r: 20})
```

Perhaps we can find a way to make both work.